// haste-sdk/AbstractHastePackage.ts

import MainWindowController from "../controllers/MainWindowController";
import {HastePackageInterface} from "./HastePackageInterface";
import HasteRowItem from "./HasteRowItem";
import Haste from "../services/Haste";
const defaultIcon = 'themes/default/images/skull.png';

class AbstractHastePackage implements HastePackageInterface
{
    protected packageData: object;
    protected packageName: string;
    protected haste: Haste;
    protected icon: string;
    constructor() {
        this.packageData = {name: this.constructor.name, path: __dirname};
        this.packageName = this.constructor.name;
        this.icon        = defaultIcon;

        /**
         * @type {Haste}
         */
        this.haste = null;
    }
    getPackageName(): string {
        return this.packageName;
    }

    insert(value, description=null, path=null, icon=null) {
        let item = new HasteRowItem();
        item.title = value;
        item.path = path ? path : value;
        item.icon = icon ? icon : this.icon;
        item.description = description ? description : "";
        this.haste.insert(item).go()
            .then(data => console.log(data))
            .catch(err => console.error(err));
    }

    search(value: string, callback: Function) {console.error('No override "search" method found in ' + this.packageName)}
    activate(rowItem: HasteRowItem, callback: Function) {console.error('No override "action" method found in ' + this.packageName)}
    //remove(rowItem: HasteRowItem, callback: Function) {console.error('No override "remove" method found in ' + this.packageName)}
}
module.exports = AbstractHastePackage;
